deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java

```diff
@@ -109,6 +110,21 @@
         new ReflectiveClassBuildItem(true, true, ThreadLocalTimestampGenerator.class.getName()));
   }
 
+  @BuildStep
+  List<ReflectiveClassBuildItem> registerLz4ForReflection(
+      CassandraClientBuildTimeConfig buildTimeConfig) {
+    if (buildTimeConfig.protocolCompression.equalsIgnoreCase("lz4")) {
+      return Arrays.asList(
+          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4Compressor"),
```

Comment:
Out of curiosity how to do establish this list? Is it by trial and error? How can we be sure that we are not forgetting a class here?

Reply:
I did it by trial and error. `Lz4` is working now (IT test is covering it) so I think there is no way that we forgot any class, right?

Reply:
I guess I'm more concerned that we might be including/referencing some things we don't need to here.

Tests still pass even if I remove one of those dependencies:

```
diff --git a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
index 70be4e2..16c920f 100644
--- a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
+++ b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
@@ -118,8 +118,8 @@ class CassandraClientProcessor {
           new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4Compressor"),
           new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaSafeCompressor"),
           new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4HCJavaSafeCompressor"),
-          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaSafeFastDecompressor"),
-          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaSafeSafeDecompressor"));
+          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaSafeFastDecompressor"));
+      // new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaSafeSafeDecompressor"));
     } else {
       return Collections.emptyList();
     }
```

Is it worth taking a look at the lz4 impl and seeing if we can deduce what classes will be needed?  Or are classes being included/used based on various configuration/runtime options which aren't being exercised here?

Reply:
Unfortunately the only way to really confirm that @tomekl007 is to look at the loading code.  It's entirely possible (arguably even likely) that the logic may take different paths based on env vars, sys props or any other test and that that logic might play differently on other systems.  That was certainly true for jnr and it looks like it's true here.... although I think what you have now will work as the base (if not optimized) case.

The driver asks the lz4 impl for the fastest instance it can come up with (https://github.com/datastax/java-driver/blob/2863100b736f01b77f7347d17536da6f151b4583/core/src/main/java/com/datastax/oss/driver/internal/core/protocol/Lz4Compressor.java#L39).  This will actually try to load a native (JNI) instance, falling back onto a pure Java instance if that fails (which it will in our case) (https://github.com/lz4/lz4-java/blob/1.6.0/src/java/net/jpountz/lz4/LZ4Factory.java#L160-L171).

There are two Java instances to choose from, one using Unsafe and the other a "safe" instance.  The loader code will try the Unsafe version first (assuming it looks like Unsafe is, er, "safe" on the current architecture) and fall back to the "safe" version if that fails (https://github.com/lz4/lz4-java/blob/1.6.0/src/java/net/jpountz/lz4/LZ4Factory.java#L136-L146).  The actual test to see if Unsafe is supported is pretty straightforward (https://github.com/lz4/lz4-java/blob/1.6.0/src/java/net/jpountz/util/Utils.java#L25-L34); it returns true for my FC28 instance and I'm guessing it returns true for you as well.  Because the "JavaUnsafe" classes _aren't_ included via the reflection config, however, you get a CNFE exception when trying to load them... which will cause the loader code to return the "JavaSafe" versions as the baseline default.  That's why you saw those classes referenced explicitly in your testing and that's why including them makes the whole thing work.

To validate this, I confirmed that native integration tests pass with the following diff applied:

```
diff --git a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
index 70be4e2..965f901 100644
--- a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
+++ b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java
@@ -116,10 +116,11 @@ class CassandraClientProcessor {
     if (buildTimeConfig.protocolCompression.equalsIgnoreCase("lz4")) {
       return Arrays.asList(
           new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4Compressor"),
-          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaSafeCompressor"),
-          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4HCJavaSafeCompressor"),
-          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaSafeFastDecompressor"),
-          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaSafeSafeDecompressor"));
+          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaUnsafeCompressor"),
+          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4HCJavaUnsafeCompressor"),
+          new ReflectiveClassBuildItem(true, true, "net.jpountz.lz4.LZ4JavaUnsafeFastDecompressor"),
+          new ReflectiveClassBuildItem(
+              true, true, "net.jpountz.lz4.LZ4JavaUnsafeSafeDecompressor"));
     } else {
       return Collections.emptyList();
     }
```

All of this leads me to one comment and one question:

* We should probably include both the "JavaSafe" and "JavaUnsafe" classes in the reflection config.  That way if a user has the ability to leverage the (faster) Unsafe impl they can do so
* The initial call from driver code will try to load a JNI impl... but we really don't want to do that in a Graal context.  Should we consider changing this behaviour (perhaps via a broader substitution class below) to just call Lz4Factory.fastestJavaInstance() in the Graal case?

Reply:
Thanks for your investigation! 
- I agree that we should provide both versions, done in https://github.com/datastax/cassandra-quarkus/pull/6/commits/192657b3c7cd548d8924a673f422f9e5985880f9
- I think that we don't need to alter (substitute) this logic. It seems that the Lz4 Loader logic is written in that way to work exactly in such cases. If it works, I think we should not add additional complexity to remove only one method call, wdyt?

Reply:
I'm fine with the change to include the unsafe variants.  And I mostly agree with you about not bothering with the shift to Lz4Factory.fastestJavaInstance() for Quarkus, at least for now.  The users that would actually pay the penalty of trying the native code first is (a) Graal users who are (b) also using compression... which can't be a huge population of users.  And those users pay a one-time cost that isn't that big anyway.  I'm not convinced it's worth the engineering effort to fix, and even if we do it should be fixed in the Java driver (using an if block based on whether or no we're running in a Graal VM)... it's not something we should fix with a substitution.